services/web/server/src/simcore_service_webserver/computation_api.py

```diff
@@ -437,6 +445,135 @@ async def update_pipeline_db(
+    if not task:
+        log.error(
+            "Task for user_id %s, project %s could not be started", user_id, project_id
+        )
+        return
+
+    log.debug(
+        "Task (task=%s, user_id=%s, project_id=%s) submitted for execution.",
+        task.task_id,
+        user_id,
+        project_id,
+    )
+    return task.task_id
+
+
+def _from_celery_state(celery_state) -> RunningState:
+    CELERY_TO_RUNNING_STATE = {
+        "PENDING": RunningState.unknown,  # TODO: Celery pending state means unknown
+        "STARTED": RunningState.started,
+        "RETRY": RunningState.retrying,
+        "FAILURE": RunningState.failure,
+        "SUCCESS": RunningState.success,
+    }
+    return RunningState(CELERY_TO_RUNNING_STATE[celery_state])
+
+
+def convert_state_from_db(db_state: StateType) -> RunningState:
+    DB_TO_RUNNING_STATE = {
+        StateType.FAILED: RunningState.failure,
+        StateType.PENDING: RunningState.pending,
+        StateType.SUCCESS: RunningState.success,
+        StateType.PUBLISHED: RunningState.published,
+        StateType.NOT_STARTED: RunningState.not_started,
+        StateType.RUNNING: RunningState.started,
+    }
+    return RunningState(DB_TO_RUNNING_STATE[StateType(db_state)])
+
+
+async def get_task_states(
+    app: web.Application, project_id: str
+) -> Dict[str, Tuple[RunningState, datetime]]:
+    db_engine = app[APP_DB_ENGINE_KEY]
+    task_states: Dict[str, RunningState] = {}
+    async with db_engine.acquire() as conn:
+        async for row in conn.execute(
+            sa.select([comp_tasks]).where(comp_tasks.c.project_id == project_id)
+        ):
+            if row.node_class != NodeClass.COMPUTATIONAL:
+                continue
+            task_states[row.node_id] = (convert_state_from_db(row.state), row.submit)
+            # the task might be running, better ask celery (NOTE this remains only 24h and disappears and state will be pending)
+            # task_result = AsyncResult(row.job_id)
+
+            # running_state = _from_celery_state(task_result.state)
+            # task_states[row.node_id] = running_state
+    return task_states
+
+
+async def get_pipeline_state(app: web.Application, project_id: str) -> RunningState:
+    task_states: Dict[str, Tuple[RunningState, datetime]] = await get_task_states(
+        app, project_id
+    )
+    # compute pipeline state from task states
+    now = datetime.utcnow()
+    if task_states:
+        # put in a set of unique values
+        set_states = {state[0] for state in task_states.values()}
+        last_update = next(
+            iter(sorted([time[1] for time in task_states.values()], reverse=True))
+        )
+        if RunningState.published in set_states:
+            if (now - last_update).seconds > get_celery_publication_timeout(app):
+                return RunningState.not_started
+        if len(set_states) == 1:
+            # this is typically for success, pending, published
+            return next(iter(set_states))
```

Comment:
MINOR: `` return set_states.pop()``  ??

Reply:
why would I do that?

Reply:
no strong reason .. i just find it more readable to get the last element of your set. It is a MINOR comment